### PR TITLE
feat(barbican): Support OpenStack Barbican authentication via Application Credentials

### PR DESCRIPTION
**Pull request by JaviMore**
## Problem Statement

What is the problem you're trying to solve?
The Barbican provider currently relies on password-based Keystone authentication. In many OpenStack environments, using user passwords for machine-to-machine access is discouraged or prohibited due to security and operational constraints (password rotation impact, broad user scope, and weaker credential isolation). This makes it harder to follow least-privilege practices and can block adoption in environments where only application credentials are allowed.

## Related Issue
https://github.com/external-secrets/external-secrets/issues/6081

## Proposed Changes

How do you like to solve the issue and why?
Adding OpenStack Application Credential authentication to the Barbican provider. The implementation extends the CRD API (secretstore_barbican_types.go) with a new authType enum (password | applicationCredential) and two new reference types for the credential ID and secret. The provider logic (provider.go) uses gophercloud/gophercloud/v2 to build AuthOptions for either auth mode, dispatched via a switch on the auth type — defaulting to password when unset, ensuring full backward compatibility.

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

Adds OpenStack Application Credential authentication to the Barbican provider while preserving password-based Keystone auth as the default.

Changes

API types
- New BarbicanAuthType enum: "password" | "applicationCredential" (default: "password").
- New reference types: BarbicanProviderAppCredIDRef and BarbicanProviderAppCredSecretRef.
- BarbicanAuth now uses optional AuthType and pointer fields for Username/Password and ApplicationCredentialID/ApplicationCredentialSecret; kubebuilder defaults and x-kubernetes-validations enforce required field groups per authType.
- Generated deepcopy methods updated/added for the new types.

Provider implementation
- Validation and client construction dispatch on authType (nil defaults to password); unknown types return errUnsupportedAuth.
- Added helpers: resolveAuthType, validatePasswordAuth, validateAppCredAuth.
- Added builders buildPasswordAuthOpts and buildAppCredAuthOpts to construct gophercloud/v2 AuthOptions, resolving secret refs and validating non-empty credentials.
- Refactored auth flow and extracted OpenStack boundary calls for testability.

CRDs / manifests
- SecretStore, ClusterSecretStore, and bundled CRDs updated with auth.applicationCredentialID, auth.applicationCredentialSecret, and auth.authType.
- Removed unconditional required username/password; added x-kubernetes-validations to require username+password when authType=password, or applicationCredentialID+applicationCredentialSecret when authType=applicationCredential, and to forbid mixing modes.

Tests
- Expanded unit tests covering defaulting (nil authType -> password), password and application-credential success/failure scenarios, inline vs secretRef handling, missing fields/authURL, and unsupported authType.
- Fixtures updated for pointer-based auth fields; OpenStack boundary functions stubbed for deterministic tests.

Docs
- API and provider docs updated with authType, new app-credential fields, examples, and usage guidance.

Backward compatibility
- Existing password-based configurations continue to work; password remains the default when authType is omitted.

### CODE REVIEW

**Comment by JaviMore** [comment]
@coderabbitai review

**Comment by JaviMore** [comment]
@coderabbitai review

**Comment by JaviMore** [comment]
@coderabbitai review

**Comment by JaviMore** [comment]
@coderabbitai review

File: deploy/crds/bundle.yaml [hunk 1]
```diff
@@ -2866,6 +2866,87 @@ spec:
                         auth:
                           description: BarbicanAuth contains the authentication information for Barbican.
                           properties:
+                            applicationCredentialID:
+                              description: ID of the application credential used for authentication.
```
[the remark this answers is not in the document]
Comment (JaviMore) [reply deploy/crds/bundle.yaml:3161]: Already addressed

**Comment by Skarlso** [comment]
Well done for keeping the initial PR small!!! I really really appreciate that.

**Comment by Skarlso** [comment]
@JaviMore sorry for the delay, there is just so much. If you would like to pick this up and update it so it has a go mod, I can review it and get this merged. :)

**Comment by JaviMore** [comment]
> @JaviMore sorry for the delay, there is just so much. If you would like to pick this up and update it so it has a go mod, I can review it and get this merged. :)

No worries at all! I've already addressed that. It's ready for another look whenever you have time. Thanks for taking the time to review! 🙂

**Comment by evrardj-roche** [comment]
@rkferreira this is the PR we mentioned in the issue :) Tell us how that goes.
Alternatively, maybe we should implement E2E test with a cloud provider credentials, but that's a longer conversation I assume :)

File: apis/externalsecrets/v1/secretstore_barbican_types.go [hunk 2]
```diff
@@ -43,7 +67,28 @@ type BarbicanProvider struct {
...
+	// - "password": use username and password.
+	// - "applicationCredential": use application credential ID and secret.
+	// Defaults to "password".
+	// +optional
+	// +kubebuilder:default="password"
+	AuthType *BarbicanAuthType `json:"authType,omitempty"`
+
+	// Username / Password authentication fields.
+	// +optional
+	Username *BarbicanProviderUsernameRef `json:"username,omitempty"`
```
Comment (evrardj-roche) [inline apis/externalsecrets/v1/secretstore_barbican_types.go:87]: should we validate with MinLength?

File: apis/externalsecrets/v1/secretstore_barbican_types.go [hunk 3]
```diff
@@ -43,7 +67,28 @@ type BarbicanProvider struct {
...
+	// +optional
+	Password *BarbicanProviderPasswordRef `json:"password,omitempty"`
```
Comment (evrardj-roche) [inline apis/externalsecrets/v1/secretstore_barbican_types.go:89]: should we validate with MinLength?

File: apis/externalsecrets/v1/secretstore_barbican_types.go [hunk 4]
```diff
@@ -33,6 +44,19 @@ type BarbicanProviderPasswordRef struct {
 	SecretRef *esmeta.SecretKeySelector `json:"secretRef"`
 }
 
+// BarbicanProviderAppCredIDRef defines a reference to an Application Credential ID.
+// +kubebuilder:validation:MinProperties=1
+// +kubebuilder:validation:MaxProperties=1
+type BarbicanProviderAppCredIDRef struct {
+	Value     string                    `json:"value,omitempty"`
```
Comment (evrardj-roche) [inline apis/externalsecrets/v1/secretstore_barbican_types.go:53]: should we validate Value (string) with MinLength?

**Review by evrardj-roche — CHANGES_REQUESTED** [review]
Please check this validation

File: apis/externalsecrets/v1/secretstore_barbican_types.go (hunk 4 above)
Comment (JaviMore) [reply apis/externalsecrets/v1/secretstore_barbican_types.go:53]: Added MinLenght validation to String fields

**Comment by evrardj-roche** [comment]
Thanks @JaviMore for updating! 

I will put this back in my queue.

**Comment by rkferreira** [comment]
I'm just running this branch, it seems not breaking anything from previous auth method. Secrets are synchronizing.

**Comment by rkferreira** [comment]
Also, tested this with App Creds, seems fine.

```
kubectl -n default create secret generic barbican-appcred \
  --from-literal=appCredId='xxx' \
  --from-literal=appCredSecret='xxx'
```


```
kubectl get SecretStore barbican-app -o yaml
apiVersion: external-secrets.io/v1
kind: SecretStore
metadata:
(...)
  name: barbican-app
  namespace: default
spec:
  provider:
    barbican:
      auth:
        applicationCredentialID:
          secretRef:
            key: appCredId
            name: barbican-appcred
        applicationCredentialSecret:
          secretRef:
            key: appCredSecret
            name: barbican-appcred
        authType: applicationCredential
      authURL: https://xxx:5000/v3
      domainName: xxx
      region: xxx
      tenantName: xxx
status:
  capabilities: ReadOnly
  conditions:
  - lastTransitionTime: "2026-08-28T20:44:33Z"
    message: store validated
    reason: Valid
    status: "True"
    type: Ready
```



```
kubectl get SecretStore
NAME           AGE    STATUS   CAPABILITIES   READY
barbican       266d   Valid    ReadOnly       True
barbican-app   55s    Valid    ReadOnly       True
```


```
3m10s       Normal    Valid                secretstore/barbican                                 store validated
42s         Normal    Valid                secretstore/barbican-app                             store validated
```



```
"level":"info","ts":1787950072.509973,"logger":"controllers.ExternalSecret","msg":"reconciled secret","ExternalSecret":{"name":"barbican-secret-app-test","namespace":"default"}}


kubectl get ExternalSecrets
NAME                                   STORETYPE     STORE          REFRESH INTERVAL   STATUS              READY   LAST SYNC
barbican-external-secret               SecretStore   barbican       0h1m0s             SecretSynced        True    8s
barbican-external-secret-02            SecretStore   barbican       1h0m0s             SecretSynced        True    45m
barbican-secret-app-test               SecretStore   barbican-app   0h1m0s             SecretSynced        True    32s
```

**Review by evrardj-roche — APPROVED** [review]

**Comment by evrardj-roche** [comment]
We have an edge case here I did not cover before, zizmor check is noop and blocks CI, no code was changed since the introduction of a new test, so this is considered as noop. Fix is to rebase. But because I did not want to wait, I will rebase it myself as I am allowed to push to your branch.

**Comment by evrardj-roche** [comment]
> We have an edge case here I did not cover before, zizmor check is noop and blocks CI, no code was changed since the introduction of a new test, so this is considered as noop. Fix is to rebase. But because I did not want to wait, I will rebase it myself as I am allowed to push to your branch.

seems it did not do the trick, I will revert and fix it differently.

**Comment by JaviMore** [comment]
@evrardj-roche Thanks for the effort

**Comment by evrardj-roche** [comment]
@JaviMore you're welcome :)
